### PR TITLE
Replacement filter

### PR DESCRIPTION
**Pull request by marcins**
Lets you restrict which files have their _contents_ checked for hashes. For example, if you know that only your "manifest" contains hashes this can significantly speed up the plugin.

Includes a safety net if this feature is used to check all files at the end for the old hashes in a single pass.

### CODE REVIEW

File: src/OutputHash.js [hunk 1]
```diff
@@ -2,9 +2,15 @@ const crypto = require('crypto');
...
-function OutputHash({ validateOutput = false, validateOutputRegex = /^.*$/ } = {}) {
+function OutputHash({
+    validateOutput = false,
+    validateOutputRegex = /^.*$/,
+    replacementFilterRegex = null,
+} = {}) {
     this.validateOutput = validateOutput;
     this.validateOutputRegex = validateOutputRegex;
+    this.replacementFilterEnabled = replacementFilterRegex !== null;
+    this.replacementFilterRegex = replacementFilterRegex || /^.*$/;
```
[the remark this answers is not in the document]
Comment (Li0liQ) [reply src/OutputHash.js]: Disregard that. `replacementFilterEnabled` is only used to perform safety-net validation at the end. Probably, renaming the variable into something like `shouldPerformReplacementValidation` could alleviate a bit of confusion.

File: src/OutputHash.js [hunk 2]
```diff
@@ -189,10 +229,33 @@ OutputHash.prototype.apply = function apply(compiler) {
...
-            replaceOldHashForNewInChunkFiles(chunk, assets, nameMap);
-            reHashChunk(chunk, assets, hashFn, nameMap);
+            replaceOldHashForNewInChunkFiles(chunk, assets, nameMap, this.replacementFilterRegex);
+            reHashChunk(chunk, assets, hashFn, nameMap, this.replacementFilterRegex);
         });
 
+        // if we are only replacing hashes in some files, we execute a safety net to ensure
+        // that we haven't missed any files that might contain hashes
+        if (this.replacementFilterEnabled) {
+            const newHashes = Object.keys(nameMap).map(key => nameMap[key]);
```
Comment (Li0liQ) [inline src/OutputHash.js]: /nit Maybe, it's worth dropping `node@6` support and use `Object.values(nameMap)`?

Comment (marcins) [reply src/OutputHash.js]: Yeah, we could do that for our fork - that's actually what I had originally until the tests failed: https://github.com/atlassian/webpack-plugin-hash-output/pull/2/commits/2d0d5e97fa00f1f778507520a3c887fdf3c358b6

Comment (marcins) [reply src/OutputHash.js]: I'll leave it for now though.

**Pull request merged**